### PR TITLE
DSL2: Adding endorspy

### PR DESCRIPTION
**Pull request by aidaanva**
Added endorspy version 1.3

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
    - [ ] If you've added a new tool - add to the software_versions process and a regex to `scrape_software_versions.py`
    - [x] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](<https://github.com/>nf-core/eager/tree/master/.github/CONTRIBUTING.md)
    - [ ] If necessary, also make a PR on the nf-core/eager _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint .`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

**Review by jbv2 — APPROVED** [review]

**Review by jfy133 — CHANGES_REQUESTED** [review]
Missing `citations.md` update (either with zenodo or eager2)

File: subworkflows/local/endorspy.nf [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+include { ENDORSPY as ENDORSPY_QUALITY_FILTER        } from '../../modules/nf-core/endorspy/main'
+include { ENDORSPY as ENDORSPY_RAW                   } from '../../modules/nf-core/endorspy/main'
+include { ENDORSPY as ENDORSPY_NOQF_DEDUP            } from '../../modules/nf-core/endorspy/main'
+
+workflow ENDORSPY {
+    take:
+    flagstats_raw // [ [meta], [flagstats_raw] ]
+    flagstats_filtered // [ [meta], [flagstats_filtered] ]
+    flagstats_duplicated // [ [meta], [flagstats_duplicated] ]
+//outputs for the flagstats will look like:     tuple val(meta), path("*.flagstat"), emit: flagstat
```
Comment (TCLamnidis) [inline subworkflows/local/endorspy.nf:15]: Harshil align

File: subworkflows/local/endorspy.nf [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+    main:
+    ch_versions       = Channel.empty()
+    ch_multiqc_files  = Channel.empty()
+
+    if ( params.run_bamfiltering & !params.skip_deduplication ) {
```
Comment (TCLamnidis) [inline subworkflows/local/endorspy.nf:21]:
```suggestion
    if ( params.run_bamfiltering && !params.skip_deduplication ) {
```

File: subworkflows/local/endorspy.nf [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+                                                .join (flagstats_duplicated)
+        ENDORSPY_COMPLETE ( ch_for_endorspy_complete )
+    } else if ( params.run_bamfiltering & params.skip_deduplication ) {
+        ch_for_endorspy_quality_filtered = flagstats_raw.join (flagstats_filtered)
+                                                        . map{
+                                                            meta, flags_raw, flags_filtered ->
+                                                            [ meta, flags_raw, flags_filtered, [] ]
+                                                            }
+        ENDORSPY_QUALITY_FILTER ( ch_for_endorspy_quality_filtered)
+    } else if ( !params.run_bamfiltering & !params.skip_deduplication) {
```
Comment (TCLamnidis) [inline subworkflows/local/endorspy.nf:32]:
```suggestion
    } else if ( !params.run_bamfiltering && !params.skip_deduplication) {
```

File: workflows/eager.nf [hunk 4]
```diff
@@ -215,6 +216,15 @@ workflow EAGER {
         ch_dedupped_flagstat = Channel.empty()
     }
 
+    //
+    // SUBWORKFLOW: calculating percent on target and clonality
+    // ENDORSPY (raw, filtered, deduplicated)
```
Comment (TCLamnidis) [inline workflows/eager.nf:221]:
```suggestion
    // ENDORSPY (raw, filtered, deduplicated)
    //
    
```

**Review by TCLamnidis — CHANGES_REQUESTED** [review]

File: subworkflows/local/endorspy.nf (hunk 1 above)
Comment (jfy133) [reply subworkflows/local/endorspy.nf:15]: tm™️

File: subworkflows/local/endorspy.nf [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    flagstats_duplicated // [ [meta], [flagstats_duplicated] ]
+//outputs for the flagstats will look like:     tuple val(meta), path("*.flagstat"), emit: flagstat
+
+    main:
+    ch_versions       = Channel.empty()
+    ch_multiqc_files  = Channel.empty()
+
+    if ( params.run_bamfiltering & !params.skip_deduplication ) {
+        ch_for_endorspy_complete = flagstats_raw.join (flagstats_filtered)
+                                                .join (flagstats_duplicated)
```
Comment (jfy133) [inline subworkflows/local/endorspy.nf:23]:
```suggestion
        ch_for_endorspy_complete = flagstats_raw
                                                                 .join (flagstats_filtered)
                                                                 .join (flagstats_duplicated)
```

File: subworkflows/local/endorspy.nf [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+        ENDORSPY_COMPLETE ( ch_for_endorspy_complete )
+    } else if ( params.run_bamfiltering & params.skip_deduplication ) {
+        ch_for_endorspy_quality_filtered = flagstats_raw.join (flagstats_filtered)
+                                                        . map{
+                                                            meta, flags_raw, flags_filtered ->
+                                                            [ meta, flags_raw, flags_filtered, [] ]
+                                                            }
```
Comment (jfy133) [inline subworkflows/local/endorspy.nf:30]:
```suggestion
                                                        .map {
                                                            meta, flags_raw, flags_filtered ->
                                                            [ meta, flags_raw, flags_filtered, [] ]
                                                        }
```

And in all others

**Review by jfy133 — COMMENTED** [review]
- Missing version mixing

File: workflows/eager.nf [hunk 7]
```diff
@@ -218,6 +220,22 @@ workflow EAGER {
...
+
+    ch_for_endorspy_filterbam = params.run_bamfiltering ? FILTER_BAM.out.flagstat : Channel.empty()
+    ch_for_endorspy_dedupped  = !params.skip_deduplication ? DEDUPLICATE.out.flagstat : Channel.empty()
+
+    //  TODO: what if not mapping executed due to bam input, figure it out later
+    ch_for_endorspy = MAP.out.flagstat
+                                    .join(ch_for_endorspy_filterbam.ifEmpty([ [], [] ]))
+                                    .join(ch_for_endorspy_dedupped.ifEmpty([ [], [] ]))
+    ENDORSPY ( ch_for_endorspy )
+
```
Comment (jfy133) [inline workflows/eager.nf:234]:
```suggestion
                                    .join(ch_for_endorspy_dedupped.ifEmpty([ [], [] ]))
                                    
    ENDORSPY ( ch_for_endorspy )
```

File: workflows/eager.nf [hunk 8]
```diff
@@ -230,6 +248,7 @@ workflow EAGER {
         ch_versions = ch_versions.mix( PRESEQ_LCEXTRAP.out.versions )
     }
 
+
```
Comment (jfy133) [inline workflows/eager.nf:251]:
```suggestion
```

File: workflows/eager.nf [hunk 9]
```diff
@@ -218,6 +220,22 @@ workflow EAGER {
...
+    //  TODO: what if not mapping executed due to bam input, figure it out later
+    ch_for_endorspy = MAP.out.flagstat
+                                    .join(ch_for_endorspy_filterbam.ifEmpty([ [], [] ]))
+                                    .join(ch_for_endorspy_dedupped.ifEmpty([ [], [] ]))
+    ENDORSPY ( ch_for_endorspy )
+
+    ch_versions   = ch_versions.mix( ENDORSPY.out.versions )
+    ch_multiqc_files  = ch_multiqc_files.mix( ENDORSPY.out.json )
+
+
```
Comment (jfy133) [inline workflows/eager.nf:238]:
```suggestion
```

**Review by jfy133 — APPROVED** [review]
Nice refactoring!

File: workflows/eager.nf [hunk 10]
```diff
@@ -218,6 +220,21 @@ workflow EAGER {
...
+    ch_for_endorspy_dedupped  = !params.skip_deduplication ? DEDUPLICATE.out.flagstat : Channel.empty()
+
+    //  TODO: what if not mapping executed due to bam input, figure it out later
+    ch_for_endorspy = MAP.out.flagstat
+                                    .join(ch_for_endorspy_filterbam.ifEmpty([ [], [] ]))
+                                    .join(ch_for_endorspy_dedupped.ifEmpty([ [], [] ]))
+                                    
+    ENDORSPY ( ch_for_endorspy )
+    ch_versions   = ch_versions.mix( ENDORSPY.out.versions )
+    ch_multiqc_files  = ch_multiqc_files.mix( ENDORSPY.out.json )
```
Comment (TCLamnidis) [inline workflows/eager.nf:236]: Harshil Align

**Review by TCLamnidis — CHANGES_REQUESTED** [review]

**Comment by aidaanva** [comment]
@nf-core bot fix linting

**Comment by TCLamnidis** [comment]
@nf-core-bot fix linting

**Review by TCLamnidis — CHANGES_REQUESTED** [review]
- [ ] Minor changes to `output.md`.

- [x] Minor changes to channel operator formatting in `eager.nf`

- [ ] Please check `modules.conf`. It seems the     `withName: 'SAMTOOLS_INDEX_MEM|SAMTOOLS_INDEX_BT2' {` block has been duplicated (probably due to a merge conflict resolve?). Please ensure module configurations are in the correct "category block" and remove the superfluous config block for `SAMTOOLS_INDEX_MEM|SAMTOOLS_INDEX_BT2`
- [x] There seems to be a superfluous `ENDORSPY.out.json` in `eager.nf`. Does that do something? if not, remove it.

File: conf/modules.config [hunk 11]
```diff
@@ -369,6 +382,17 @@ process {
         ]
     }
 
+
```
Comment (TCLamnidis) [inline conf/modules.config:385]:
```suggestion
```

File: conf/modules.config [hunk 12]
```diff
@@ -258,6 +258,19 @@ process {
         ]
     }
 
+    //
+    // BAM INPUT
+    //
+    withName: 'SAMTOOLS_FLAGSTATS_BAM_INPUT' {
+        tag = { "${meta.reference}|${meta.sample_id}_${meta.library_id}_L${meta.lane}" }
```
Comment (TCLamnidis) [inline conf/modules.config:265]:
```suggestion
        // TODO Once a lane-merging step is added for input BAMs, the lane should be dropped from this tag.
        tag = { "${meta.reference}|${meta.sample_id}_${meta.library_id}_L${meta.lane}" }
```

File: docs/output.md [hunk 13]
```diff
@@ -199,6 +199,76 @@ The resulting FASTQ files can be used for data sharing purposes where the host D
...
+
+Additionally, this script provides two different ways of estimating library complexity:
+
+Clonality (Cluster Factor in eager1): ratio of reads that have duplicated reads
+
+$$ Clonality = { TotalReadsPreDedup \over mappedDedup } $$
+
+Percent Duplicates (%): percent of mapping reads that have at least 1 duplicate. It is calculated as follows:
+
+$$ Percent_Duplicates = {(TotalReadsPreDedup - MappedReadsDedup) \over TotalReadsPreDedup \* 100} $$
```
Comment (TCLamnidis) [inline docs/output.md:247]:
```suggestion
$$ PercentDuplicates = {(TotalReadsPreDedup - MappedReadsDedup) \over TotalReadsPreDedup \* 100} $$
```

**Review by TCLamnidis — APPROVED** [review]
Minor tweaks. Added a future TODO.
Otherwise LGTM!

**Pull request merged**